File: dsa/sorting/merge_sort.py
def merge_sort(arr):
    """
    Args:
     arr(list_int32)
    Returns:
     list_int32
    """
    # Write your code here.
    # create helper fn for recursion
    def merger(arr):
        # set base case: if len(arr) <= 1, return
        if(len(arr) <= 1):
            return arr
        # get midpoint
        mid = len(arr) // 2
        # get left partition
        left = arr[:mid]
        # get right partition
        right = arr[mid:]
        # call helper fn on left partition
        merger(left)
        # call helper fn on right partition
        merger(right)
        # compare values from left and right partitions to sort
        [leftInd, rightInd, finalInd] = [0, 0, 0]
        while ( (leftInd < len(left)) & (rightInd < len(right)) ):
            if(left[leftInd] < right[rightInd]):
                arr[finalInd] = left[leftInd]
                leftInd += 1
                finalInd += 1
            else:
                arr[finalInd] = right[rightInd]
                rightInd += 1
                finalInd += 1
        # empty contents of remaining arr into final arr
        while (leftInd < len(left)):
            arr[finalInd] = left[leftInd]
            leftInd += 1
            finalInd += 1
        while(rightInd < len(right)):
            arr[finalInd] = right[rightInd]
            rightInd += 1
            finalInd += 1
        # return sorted arr
        return arr
    return merger(arr)

File: dsa/cpp/algorithms/sort/mergeSort.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> merge_sort(vector<int> &v) {
    // make merge helper fn
    // return v if its length <= 1
    if (v.size() <= 1) {
        return v;
    }
    // get mid index
    int midInd = v.size()/2;
    // get left partition
    vector<int> left = vector<int>(v.begin(), v.begin()+midInd);
    // get right partition
    vector<int> right = vector<int>(v.begin()+midInd,v.end());
    // call helper fn on left partition
    merge_sort(left);
    // call helper fn on right partition
    merge_sort(right);
    // compare values from left and right partitions to sort
    int leftInd, rightInd, finalInd;
    leftInd = rightInd = finalInd = 0;

    while ( (leftInd < left.size()) && (rightInd < right.size()) ) {
        if (left[leftInd] < right[rightInd]) {
            v[finalInd] = left[leftInd];
            leftInd += 1;
        } else {
            v[finalInd] = right[rightInd];
            rightInd += 1;
        }
        finalInd += 1;
    }
    // empty contents of remaining partition into final arr
    while (leftInd < left.size()) {
        v[finalInd] = left[leftInd];
        leftInd += 1;
        finalInd += 1;
    }
    while (rightInd < right.size()) {
        v[finalInd] = right[rightInd];
        rightInd += 1;
        finalInd += 1;
    }

    // return sorted arr
    return v;
}

int main_merge()
{
    vector<int> sample_v = {6, 34, 24, 32, 23};
    vector<int> sorted_v = merge_sort(sample_v);
    for (int num : sorted_v)
    {
        cout << num << " " << endl;
    }
    return 0;
}

